docs/api.rst

```diff
@@ -0,0 +1,90 @@
+===========
+Treelib API
+===========
+
+Library for manipulating trees in Python made up of dicts and lists.
+
+
+.. py:func:: tree_get(tree, key, default=None)
+
+   Given a tree consisting of mappings (like dict) and indexable sequences (like
+   list), returns the value specified by the key.
+
+   The key is a period-delimited sequence of edges to traverse. If one of the
+   ediges doesn't exist, then the default is returned immediately.
+
+   Examples:
+
+   >>> tree_get({'a': 1}, 'a')
+   1
+   >>> tree_get({'a': 1}, 'b')
+   None
+   >>> tree_get({'a': {'b': 2}}, 'a.b')
+   2
+   >>> tree_get({'a': {'b': 2}}, 'a.b.c', default=55)
+   55
+
+   This supports sequences, too:
+
+   >>> tree_get({'a': [1, 2, 3]}, 'a.1')
+   2
+   >>> tree_get({'a': {'1': 2}}, 'a.1')
+   2
+
+   Both dict and list support getitem notation, so the ``1`` works fine.
+
+   Some things to know about ``tree_get()``:
+
+   1. It doesn't alter the tree at all.
+   2. Once it hits an edge that's missing, it returns ``None`` or the default.
+
+
+.. py:func:: tree_set(tree, key, value)
+
+   Given a tree consisting of mappings (like dict) and indexable sequences that
+   support getitem notation, sets the key to the value.
+
+   The key is a period-delimited sequence of edges to traverse. If one of the
+   ediges doesn't exist, then the edge is created using these rules:
+
+   1. if the next edge is an integer, then it creates a list
+   2. if the next edge is not an integer, then it creates a dict
+
+   This returns the tree which is mutated in place.
+
+   >>> tree_set({}, 'a', value=5)
+   {'a': 5}
+   >>> tree_set({}, 'a.b.c', value=5)
+   {'a': {'b': {'c': 5}}}
+
+   While ``tree_set`` does create new dicts and lists if they're missing, it
+   will not create new list indexes. Instead, it'll raise an ``IndexError``. For
+   example:
+
+   >>> tree_set({}, 'a.1', value=5)
+   IndexError('list index out of range')
+
+   This is the same error you'd get if you tried to access an index that doesn't
+   exist in a list.
+
```

Comment:
I had another idea that has two parts.

1. we make `tree_set` only builds dicts for things that are missing and in this way, builds "sparse lists"
2. we add an argument to `tree_set` for "don't build things that don't exist"

So then we have use cases:

**Developer wants to add something deep in the tree that doesn't involve lists.** Works as advertised.

```
>>> tree_set({}, 'a.b.c', value='foo')
{'a': {'b': {'c': 'foo'}}}
```

Done.

**Developer wants to add something deep in the tree that has a list in the middle.** If the list can be sparse (this really depends on how the tree is used), then it treats everything as a dict key and creates dicts as it goes along:

```
>>> tree_set({}, 'a.1.b', value='foo')
{'a': {'1': {'b': 'foo'}}}
```

Done.

**Developer wants lists to be lists.** At this point, the developer has to build the list manually. Python has a `setdefault`--maybe we could do something like that which creates something if it doesn't exist already.

`tree_setdefault(tree, default_tree)`

```
>>> ret = tree_setdefault({}, {'a': [None, None, None]})
>>> ret
{'a': [None, None, None]}
>>> tree_set(ret, 'a.1.b', value='foo')
{'a': [None, {'b': 'foo'}, None]}
```

That `tree_setdefault` would also let us take a tree of data and add default values to it which I think would also be super handy. There are some behavior things there we'd have to figure out like what happens if an edge in the tree leads to a different kind of node than the equivalent edge of the default tree (for example, dict vs. list).